README.md
# Setup a private obfs4 Tor bridge on Ubuntu server

Copy script to your server:

```bash
wget https://raw.githubusercontent.com/Xaqron/tor-private-bridge/master/tor-bridge.sh
```

## Easy installation

Run the script as root:

```bash
sudo bash tor-bridge.sh
```

Done :blush:

## Advanced installation

```bash
sudo bash tor-bridge.sh 172.16.81.23 51452 59009
```

### In above example

* `172.16.81.23` is your server public IP address.
* `51452` is the listening port of tor service.
* `59009` is ORPort.

For more security change `listening port` and `ORPort` to a random number of your own.

When setup finished you will get an obfs4 address which can be used in Tor browser.

Also your `obfs4` address will be saved at `obfs4.address` file in your home directory.

## Tor Client setup

Use `obfs4` address as `custom bridge` in your Tor client:

![tor-browser](images/tor-browser.png)

> Censorship reflects a society's lack of confidence in itself.


CHANGES.md
# Release notes

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org).

## 1.1.0

- Easy installation mode added (default).
- Support for all Ubuntu server versions.

## 1.0.0

- First release.


ubuntu-server-16.04.3-LTS.sh
#!/usr/bin/env bash

#=================================================
#title           :Tor bridge installer
#description     :Installs a private Tor bridge
#author	    	 :Xaqron
#date            :20180416
#version         :1.0.0
#=================================================

echo Checking root privilege...
if [[ $EUID -ne 0 ]]; then
    echo Run script as root. Exiting...
    exit 1
fi

echo Usage: command IP ListenPort ORPort

echo Updating server
apt update
apt full-upgrade -y

echo Adding official tor repositories...
rm -rf /etc/apt/sources.list.d/tor.list 2> /dev/null
echo "deb http://deb.torproject.org/torproject.org xenial main" >> /etc/apt/sources.list.d/tor.list
echo "deb-src http://deb.torproject.org/torproject.org xenial main" >> /etc/apt/sources.list.d/tor.list

echo Adding gpg keys...
apt install -y dirmngr
gpg --keyserver keys.gnupg.net --recv A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89
gpg --export A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89 | sudo apt-key add -

echo Installing Tor bridge...
apt update
apt install -y tor tor-geoipdb deb.torproject.org-keyring obfs4proxy
service tor stop

echo A backup made as /etc/tor/torrc.bak, Writing Tor config...
mv /etc/tor/torrc /etc/tor/torrc.bak
echo "RunAsDaemon 0" >> /etc/tor/torrc
echo "ORPort $3" >> /etc/tor/torrc
echo "ExtORPort auto" >> /etc/tor/torrc
echo "ExitPolicy reject *:*" >> /etc/tor/torrc
echo "BridgeRelay 1" >> /etc/tor/torrc
echo "PublishServerDescriptor 0" >> /etc/tor/torrc
echo "ServerTransportPlugin obfs4 exec /usr/bin/obfs4proxy" >> /etc/tor/torrc
echo "ServerTransportListenAddr obfs4 0.0.0.0:$2" >> /etc/tor/torrc
echo "ContactInfo https://twitter.com/xaqron" >> /etc/tor/torrc
echo "Nickname Xaqron" >> /etc/tor/torrc

echo Setting firewall rules...
iptables -A INPUT -p tcp --dport $2 -j ACCEPT
iptables -A INPUT -p tcp --dport $3 -j ACCEPT
dpkg-reconfigure iptables-persistent

echo Starting Tor service...
service tor start

echo Tor starts after 30 seconds. Please wait...
sleep 30

echo Your obfs4 address is:
OBFS4TEMPLATE=`tail -1 /var/lib/tor/pt_state/obfs4_bridgeline.txt`
FINGERPRINT0=`cat /var/lib/tor/fingerprint`
FINGERPRINT=${FINGERPRINT0#"Xaqron "}
OBFS4ADDRESS="${OBFS4TEMPLATE/<IP ADDRESS>/$1}"
OBFS4ADDRESS="${OBFS4ADDRESS/<PORT>/$2}"
OBFS4ADDRESS="${OBFS4ADDRESS/<FINGERPRINT>/$FINGERPRINT}"
echo $OBFS4ADDRESS

echo Saving obfs4 address to file: ~/obfs4.address
echo Use obfs4 address with Tor client.
echo $OBFS4ADDRESS > ~/obfs4.address
echo Done!

tor-bridge.sh
#!/usr/bin/env bash

#=================================================
#title           :Tor bridge installer
#description     :Installs a private Tor bridge
#author	    	 :Xaqron
#date            :20180416
#version         :1.1.0
#license         :MIT    
#=================================================

TOR_PUBLIC_KEY="A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89"

echo "Checking root privilege..."
if [[ $EUID -ne 0 ]]; then
    echo "Usage: sudo bash $0"
    echo "Run script as root. Exiting..."
    exit 1
fi

if [ "$#" -eq 3 ]; then # TODO: Argument validation
    IP=$1
    LISTENING_PORT=$2
    OR_PORT=$3
elif [ "$#" -eq 0 ]; then
    IP=$(curl ipinfo.io/ip)
    LISTENING_PORT=$((20000 + RANDOM % 39000))
    OR_PORT=$((20000 + RANDOM % 39000))
else
    echo "Wrong parameters!"
    echo "Usage: sudo bash $0 IP ListenPort ORPort"
    exit 1
fi

echo "Updating server..."
apt update
apt full-upgrade -y

echo "Adding gpg keys..."
apt install -y dirmngr
gpg --keyserver keys.gnupg.net --recv $TOR_PUBLIC_KEY
gpg --export $TOR_PUBLIC_KEY | apt-key add -

echo Adding official Tor repositories...
rm -f /etc/apt/sources.list.d/tor.list 2> /dev/null
rm -f /etc/apt/sources.list.d/tor.list.save 2> /dev/null
echo "deb http://deb.torproject.org/torproject.org $(lsb_release -cs) main" > /etc/apt/sources.list.d/tor.list
echo "deb-src http://deb.torproject.org/torproject.org $(lsb_release -cs) main" >> /etc/apt/sources.list.d/tor.list

echo Installing Tor bridge...
apt update
apt install -y tor tor-geoipdb deb.torproject.org-keyring obfs4proxy
service tor stop

echo "Backup made at \"/etc/tor/torrc.bak\", Writing Tor config..."
mv /etc/tor/torrc /etc/tor/torrc.bak
echo "RunAsDaemon 0" >> /etc/tor/torrc
echo "ORPort $OR_PORT" >> /etc/tor/torrc
echo "ExtORPort auto" >> /etc/tor/torrc
echo "ExitPolicy reject *:*" >> /etc/tor/torrc
echo "BridgeRelay 1" >> /etc/tor/torrc
echo "PublishServerDescriptor 0" >> /etc/tor/torrc
echo "ServerTransportPlugin obfs4 exec /usr/bin/obfs4proxy" >> /etc/tor/torrc
echo "ServerTransportListenAddr obfs4 0.0.0.0:$LISTENING_PORT" >> /etc/tor/torrc

echo "Setting firewall rules..."
ufw allow $LISTENING_PORT
ufw allow $OR_PORT

echo "Starting Tor service..."
service tor start

echo "Tor starts after 30 seconds. Please wait..."
sleep 30

echo
echo "Your obfs4 address is:"
OBFS4TEMPLATE=`tail -1 /var/lib/tor/pt_state/obfs4_bridgeline.txt`
FINGERPRINT0=`cat /var/lib/tor/fingerprint`
FINGERPRINT0=$(echo $FINGERPRINT0 | cut -d ' ' -f 2-)
FINGERPRINT=${FINGERPRINT0#"Xaqron "}
OBFS4ADDRESS="${OBFS4TEMPLATE/<IP ADDRESS>/$IP}"
OBFS4ADDRESS="${OBFS4ADDRESS/<PORT>/$LISTENING_PORT}"
OBFS4ADDRESS="${OBFS4ADDRESS/<FINGERPRINT>/$FINGERPRINT}"
echo $OBFS4ADDRESS
echo

echo "Saving obfs4 address to: \"~/obfs4.address\""
echo "Use obfs4 address with Tor client."
echo $OBFS4ADDRESS > ~/obfs4.address
echo "Done!"
